### DISCUSSION TITLE
View disappear action is not triggered when the view is disappeared

### DISCUSSION THREAD

**Discussion by johnqiang-cao** [pointfreeco/swift-composable-architecture#3520] [General]
### Description

We have code in the view:
.onDisappear {
     store.send(.viewDisappeared)
}
But viewDisappeared action is not triggered when the view is unloaded.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The .viewDisappeared action should be triggered.

### Actual behavior

The .viewDisappeared action is not triggered.

### Reproducing project

[ResourceCleanup 2.zip](https://github.com/user-attachments/files/17999122/ResourceCleanup.2.zip)

### The Composable Architecture version information

1.16.1

### Destination operating system

iOS 18

### Xcode version information

XCode 16.1

**Comment by mbrandonw** [comment]
Hi @johnqiang-cao, this is to be expected, and there is even a purple runtime displaying in Xcode letting you know what you are doing isn't correct:

> 🟣 A "forEach" at "ComposableArchitecture/CaseReducer.swift:52" received an action for a missing element. …
>
>  Action:
>    ContentFeature.Path.Action.child(.viewDisappeared)
>
> This is generally considered an application logic error, and can happen for a few reasons:

You cannot send an action in the `onDisappear` of a view because by the time that view modifier is invoked the feature's state has been `nil`'d out, and hence the reducer for the child feature cannot even be run.

What is it you are wanting to do in the `onDisappear` of the view?

If you really need to do this, you can always add it inside the `navigationDestination`:

```swift
} destination: { destination in
  switch destination.case {
  case .child(let childStore):
    ChildView(store: childStore)
      .onDisappear { store.send(.onChildDisappear…) }
  }
}
```

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Comment by johnqiang-cao** [comment]
Hi @mbrandonw , thanks for response. 
In the ChildFeature, we do have two state properties which are stuck in memory and cause memory leak if not clean:
```
@Reducer
struct ChildFeature: Sendable {
    @ObservableState
    struct State: ReducerState {
        var childModel = [ChildModel?](repeating: nil, count: 6)
        var customScenes = [ViewType: ImagingScene]()
        var imagingCache: ImagingCache = .init()
    } 
```

In Reducer body, we have action:
```
        case .viewDisappeared:
                state.boneModel = [ChildModel?](repeating: nil, count: 6)
                state.customScenes = [:]
                return .run { [state] _ in
                    await state.imagingCache.purgeLoaderCache()
               }
```

When the Child view is unloaded, from ChildView's .onDisappear, call `store.send(.viewDisappeared)` to avoid memory leak.

I do not think I can move them up to navigationDestination (with .onChildDisappear) since `ChildModel` and `customScenes` are sit inside ChildFeature.

Previously, above .`viewDisappeard` action was triggered and memory was clean. Recently we upgraded to XCode 16.1 and then found out .`viewDisappeard` action is not triggered anymore.

Any suggestion? Thanks.

**Comment by Toof-X** [comment]
I have the same issue:

On View:
```
.onDisappear {
       store.send(.disappear)  // breakpoint is performed
}
```

On Store:
```
case .disappear:
       // breakpoint is not performed
}
```

I still receive onDisappear event on View, but on Store, I can not trigger disappear action. If you have any solution, please notify me, thanks

Xcode 16.3
TCA 1.19.0

**Reply by mbrandonw** [reply]
Hi @Toof-X, did you read the above [message](https://github.com/pointfreeco/swift-composable-architecture/discussions/3520#discussioncomment-11453952)? If your view is being presented via optional state, then sending an action in `onDisappear` can't possible work because the state would have been `nil`'d out, and hence the reducer cannot be invoked.

**Reply by Toof-X** [reply]
I want to pause the Audio Player after navigation back to root, I will try to pause the audio in Navigation Reducer, thanks @mbrandonw

**Reply by mbrandonw** [reply]
That particular situation may even be easier though. If your audio player is started in an effect and respects cooperative cancellation (i.e. it stops when it detects cancellation), then everything can just work automatically (as long as you are using our presentation tools).
